tests/test_vqe.py

```diff
@@ -582,6 +716,152 @@ def test_passing_kwargs(self, coeffs, observables, expected):
             assert qnode.h == 123
             assert qnode.order == 2
 
+    def test_optimize_outside_tape_mode(self):
+        """Test that an error is raised if observable optimization is requested outside of tape
+        mode."""
+        if qml.tape_mode_active():
+            pytest.skip("This test works with tape mode disabled")
+
+        dev = qml.device("default.qubit", wires=2)
+        hamiltonian = qml.vqe.Hamiltonian([1], [qml.PauliZ(0)])
+
+        with pytest.raises(ValueError, match="Observable optimization is only supported in tape"):
+            qml.VQECost(lambda params, **kwargs: None, hamiltonian, dev, optimize=True)
+
+    @pytest.mark.parametrize("interface", ["tf", "torch", "autograd"])
+    def test_optimize(self, interface, tf_support, torch_support):
+        """Test that a VQECost with observable optimization gives the same result as another
+        VQECost without observable optimization."""
+        if not qml.tape_mode_active():
+            pytest.skip("This test works with tape mode enabled")
+        if interface == "tf" and not tf_support:
+            pytest.skip("This test requires TensorFlow")
+        if interface == "torch" and not torch_support:
+            pytest.skip("This test requires Torch")
+
+        dev = qml.device("default.qubit", wires=4)
+        hamiltonian = big_hamiltonian
+
+        cost = qml.VQECost(
+            qml.templates.StronglyEntanglingLayers,
+            hamiltonian,
+            dev,
+            optimize=True,
+            interface=interface,
+        )
+        cost2 = qml.VQECost(
+            qml.templates.StronglyEntanglingLayers,
+            hamiltonian,
+            dev,
+            optimize=False,
+            interface=interface,
+        )
+
+        w = qml.init.strong_ent_layers_uniform(2, 4, seed=1967)
+
+        c1 = cost(w)
+        exec_opt = dev.num_executions
+        dev._num_executions = 0
+
+        c2 = cost2(w)
+        exec_no_opt = dev.num_executions
+
+        assert exec_opt == 5
```

Comment:
is this known in advance due to a property of the Hamiltonian? Looks a bit like a magic number with no context 🤔

Reply:
Ah yes this is just a hard-coded thing that happens to be the number of groups in that Hamiltonian